ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java

```diff
@@ -313,14 +302,27 @@ private void updateOperationState(ReplicaId replica, RouterErrorCode error) {
   private void checkAndMaybeComplete() {
     // operationCompleted is true if Blob_Authorization_Failure was received.
     if (operationTracker.isDone() || operationCompleted) {
-      if (!operationTracker.hasSucceeded()) {
-        setOperationException(
-            new RouterException("The TtlUpdateOperation could not be completed.", resolvedRouterErrorCode));
+      if (operationTracker.hasSucceeded()) {
+        operationException.set(null);
```

Comment:
It doesn't default to null?  It's possible to not be null at this point?

Reply:
it's defaulted to be null, but it's also possible to be not null at this point. we will send three requests to ambry-server and every error response for those requests will set the operationException to something not null.